### PR TITLE
[env] Implement all methods of env_flink

### PR DESCRIPTION
**Pull request by masteryhx**
## What is the purpose of the change

Implemet all methods of env_flink

## Brief change log

  - re-construct jni_helper
  - Implement all methods of env_flink

## Verifying this change

This change will be tested in Java side after FlinkEnv is ready.

### CODE REVIEW

File: env/flink/env_flink.cc [hunk 1]
```diff
@@ -1,10 +1,844 @@
...
+// to access Flink FileSystem. All files created by one instance of ForSt
+// will reside on the actual Flink FileSystem.
+//
+namespace ROCKSDB_NAMESPACE {
+
+// Appends to an existing file in Flink FileSystem.
+class FlinkWritableFile : public FSWritableFile {
+ private:
+  std::string file_path_;
+  jobject file_system_instance_;
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: nit: file_path_ and "file_system_instance" could be modified by "const"?

File: env/flink/env_flink.cc [hunk 2]
```diff
@@ -1,10 +1,844 @@
...
+  }
+};
+
+// Used for reading a file from Flink FileSystem. It implements both
+// sequential-read access methods and random read access methods.
+class FlinkReadableFile : virtual public FSSequentialFile,
+                          virtual public FSRandomAccessFile {
+ private:
+  std::string file_path_;
+  jobject file_system_instance_;
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: nit: file_path_ and "file_system_instance" could be modified by "const"?

File: env/flink/env_flink.cc [hunk 3]
```diff
@@ -1,10 +1,844 @@
...
+
+IOStatus FlinkFileSystem::NewDirectory(const std::string& name,
+                                       const IOOptions& options,
+                                       std::unique_ptr<FSDirectory>* result,
+                                       IODebugContext* dbg) {
+  IOStatus s = FileExists(name, options, dbg);
+  if (s.ok()) {
+    result->reset(new FlinkDirectory());
+  }
+  return s;
```
Comment (ljz2051) [inline env/flink/env_flink.cc:444]: This method does not follow the interface definition ： ”On failure stores nullptr in *result and
 returns non-OK.“   I think we could call result->reset() at the beginning of this method.

File: env/flink/env_flink.cc [hunk 4]
```diff
@@ -1,10 +1,844 @@
...
+  return IOStatus::OK();
+}
+
+IOStatus FlinkFileSystem::NewDirectory(const std::string& name,
+                                       const IOOptions& options,
+                                       std::unique_ptr<FSDirectory>* result,
+                                       IODebugContext* dbg) {
+  IOStatus s = FileExists(name, options, dbg);
+  if (s.ok()) {
+    result->reset(new FlinkDirectory());
```
Comment (ljz2051) [inline env/flink/env_flink.cc:442]: How about saving the name string into the FlinkDirectory object?

File: env/flink/env_flink.cc [hunk 5]
```diff
@@ -1,10 +1,844 @@
...
+      class_cache_->ConstructPathInstance(filePath, &pathInstance);
+  if (!status.ok()) {
+    return status;
+  }
+
+  // Call delete method
+  JNIEnv* jniEnv = getJNIEnv();
+  JavaClassCache::JavaMethodContext deleteMethod =
+      class_cache_->GetJMethod(JavaClassCache::JM_FLINK_FILE_SYSTEM_DELETE);
+  jboolean created = jniEnv->CallBooleanMethod(
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: created => deleted?

File: env/flink/env_flink.cc [hunk 6]
```diff
@@ -1,10 +1,844 @@
...
+  jniEnv->DeleteLocalRef(pathInstance);
+
+  status = CurrentStatus([filePath]() {
+    return std::string("Exception when Delete, path: ").append(filePath);
+  });
+  if (!status.ok()) {
+    return status;
+  }
+
+  return created ? IOStatus::OK() : IOStatus::IOError(status.ToString());
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: The "status.ToString" here always  return "OK?"  Maybe we need to construct an error message manually.

File: env/flink/env_flink.cc [hunk 7]
```diff
@@ -1,10 +1,844 @@
...
+  });
+  if (!status.ok()) {
+    return status;
+  }
+
+  *size = fileSize;
+  return IOStatus::OK();
+}
+
+// 1. Must guarantee that FileStatus::getPath will not be used!
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: Can you explain why it must be guaranteed that FileStatus::getPath cannot be used here?

File: env/flink/env_flink.cc [hunk 8]
```diff
@@ -1,10 +1,844 @@
...
+
+  std::string srcFilePath = ConstructPath(src);
+  // Construct src Path Instance
+  jobject srcPathInstance;
+  status = class_cache_->ConstructPathInstance(srcFilePath, &srcPathInstance);
+  if (!status.ok()) {
+    return status;
+  }
+
+  std::string targetFilePath = ConstructPath(src);
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: ConstructPath(src)   =>  ConstructPath(target);

File: env/flink/env_flink.cc [hunk 9]
```diff
@@ -1,10 +1,844 @@
...
+    return std::string("Exception when RenameFile, src: ")
+        .append(srcFilePath)
+        .append(", target: ")
+        .append(targetFilePath);
+  });
+  if (!status.ok()) {
+    return status;
+  }
+
+  return renamed ? IOStatus::OK() : IOStatus::IOError(status.ToString());
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: The references of srcPathInstance and targetPathInstance are not released in this method.

**Review by ljz2051 — CHANGES_REQUESTED** [review]
Thanks for the PR!  I left some comments. PTAL

File: env/flink/env_flink.cc (hunk 7 above)
Comment (masteryhx) [reply env/flink/env_flink.cc]: Thanks for the reminder.
It works well after I reconstruct the logic, so I just removed the comment.

File: env/flink/env_flink.cc (hunk 4 above)
Comment (masteryhx) [reply env/flink/env_flink.cc:442]: This is a dummy class which shouldn't affect the normal usage in the flink side.
I just added TODO here.

File: env/flink/env_flink.h [hunk 10]
```diff
@@ -28,16 +42,10 @@ class FlinkFileSystem : public FileSystemWrapper {
   static const char* kNickName() { return "flink"; }
   const char* NickName() const override { return kNickName(); }
 
-  // Constructor and Destructor
-  explicit FlinkFileSystem(const std::shared_ptr<FileSystem>& base,
-                           const std::string& fsname);
+  // Destructor
```
Comment (ljz2051) [inline env/flink/env_flink.h]: This comment is meaningless; maybe we could omit such comments?

File: env/flink/env_flink.h [hunk 11]
```diff
@@ -90,7 +99,24 @@ class FlinkFileSystem : public FileSystemWrapper {
                        IODebugContext* /*dbg*/) override;
 
  private:
-  std::string base_path_;
+  const std::string base_path_;
+  JavaClassCache* class_cache_;
+  jobject file_system_instance_;
+
+  // Constructor and Destructor
```
Comment (ljz2051) [inline env/flink/env_flink.h]: omit such comments?

File: env/flink/jni_helper.cc [hunk 12]
```diff
@@ -1,76 +1,337 @@
...
+    if (item.javaClass) {
+      jni_env_->DeleteGlobalRef(item.javaClass);
+    }
+  }
+}
+
+IOStatus JavaClassCache::Create(JNIEnv* env,
+                                std::unique_ptr<JavaClassCache>* result) {
+  auto classCache = new JavaClassCache(env);
+  IOStatus status = classCache->Init();
```
Comment (ljz2051) [inline env/flink/jni_helper.cc:39]: If the classCache init failed,  the references of classCache should be released?

File: env/flink/jni_helper.cc [hunk 13]
```diff
@@ -1,76 +1,337 @@
...
+      .methodName = "readFully";
+  cached_java_methods_[CachedJavaMethod::JM_FLINK_FS_INPUT_STREAM_SEQ_READ]
+      .signature = "(Ljava/nio/ByteBuffer;)I";
+
+  cached_java_methods_[CachedJavaMethod::JM_FLINK_FS_INPUT_STREAM_RANDOM_READ]
+      .javaClassAndName = cached_java_classes_[JC_FLINK_FS_INPUT_STREAM];
+  cached_java_methods_[CachedJavaMethod::JM_FLINK_FS_INPUT_STREAM_RANDOM_READ]
+      .methodName = "readFully";
+  cached_java_methods_[CachedJavaMethod::JM_FLINK_FS_INPUT_STREAM_RANDOM_READ]
+      .signature = "(JLjava/nio/ByteBuffer;)I";
```
Comment (ljz2051) [inline env/flink/jni_helper.cc:172]: The signature of JM_FLINK_FS_INPUT_STREAM_RANDOM_READ should contains "position"?

File: env/flink/env_flink.cc [hunk 14]
```diff
@@ -1,10 +1,854 @@
...
+  if (!status.ok()) {
+    return status;
+  }
+
+  std::string targetFilePath = ConstructPath(target);
+  // Construct target Path Instance
+  jobject targetPathInstance;
+  status =
+      class_cache_->ConstructPathInstance(targetFilePath, &targetPathInstance);
+  if (!status.ok()) {
```
Comment (ljz2051) [inline env/flink/env_flink.cc:802]: The reference of srcPathInstance should be released here?

File: env/flink/env_flink.cc [hunk 15]
```diff
@@ -1,10 +1,854 @@
...
+    return std::string("Exception when RenameFile, src: ")
+        .append(srcFilePath)
+        .append(", target: ")
+        .append(targetFilePath);
+  });
+  if (!status.ok()) {
+    return status;
+  }
+
+  return renamed ? IOStatus::OK() : IOStatus::IOError(status.ToString());
```
Comment (ljz2051) [inline env/flink/env_flink.cc]: The "status.ToString" always return "OK". Maybe we also need to construct the error message manually here.

**Review by ljz2051 — COMMENTED** [review]
Thanks for the update. PTAL my comments.

File: env/flink/jni_helper.cc (hunk 13 above)
Comment (masteryhx) [reply env/flink/jni_helper.cc:172]: The signature has contained "J" as type of "position".

**Review by ljz2051 — APPROVED** [review]
Overall, LGTM.
As we discussed offline, we should add necessary UT tests for these filesystem interface later.

**Pull request merged**